feat: user-role assignment and search support in client

Oleksiivanov opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)
  - [ ] enable backports [when recommended](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)

## Related issues

closes #

remcowesterhoud commented on clients/java/src/main/java/io/camunda/client/CamundaClient.java:
```diff
@@ -1317,6 +1320,57 @@ ActivateAdHocSubProcessActivitiesCommandStep1 newActivateAdHocSubProcessActiviti
+   * @param roleId the ID of the role
+   * @return a builder for the assign user to role command
+   */
+  AssignUserToRoleCommandStep1 newAssignUserToRoleCommand(String roleId);
```
❌ In slack we had a [long discussion](https://camunda.slack.com/archives/C06UYJMMETZ/p1747317782261729?thread_ts=1747315273.504759&cid=C06UYJMMETZ) regarding naming yesterday. The outcome is that this command should look like:

```java
 camundaClient
    .newAssignRoleToUserCommand()
    .roleId("roleId")
    .username("username")
    .send();
```

[Here](https://github.com/camunda/camunda/pull/32053) is a good example for mappings, which are pretty much identical to users.

remcowesterhoud commented on clients/java/src/main/java/io/camunda/client/CamundaClient.java:
```diff
@@ -1317,6 +1320,57 @@ ActivateAdHocSubProcessActivitiesCommandStep1 newActivateAdHocSubProcessActiviti
+   * @param roleId the ID of the role
+   * @return a builder for the unassign user from role command
+   */
+  UnassignUserFromRoleCommandStep1 newUnassignUserFromRoleCommand(String roleId);
```
❌ Same thing here. This should look like:

```java
 camundaClient
    .newUnassignRoleFromUserCommand()
    .roleId("roleId")
    .username("username")
    .send();
```

remcowesterhoud commented on clients/java/src/main/java/io/camunda/client/api/command/AssignUserToRoleCommandStep1.java:
```diff
@@ -0,0 +1,30 @@
+
+import io.camunda.client.api.response.AssignUserToRoleResponse;
+
+public interface AssignUserToRoleCommandStep1 extends FinalCommandStep<AssignUserToRoleResponse> {
```
❌ Should become `AssignRoleToUserCommandStep1` which has a method `roleId` and returns an `AssignRoleToUserCommandStep2` which extends the `FinalCommandStep` and has a method `username`

remcowesterhoud reviewed (commented):
❌ Please add a client tests similar to the `UnassignRoleFromMappingTest` in [this PR](https://github.com/camunda/camunda/pull/32053/files#diff-6bc4418540be41dc37206311db1913243156d0423a89d9fc6d8e25ab10a99486). This is a bit more of a lightweight test that ensures the client builder creates the right request and makes it to the correct endpoint. The ITs cover it but are quite slow.

remcowesterhoud commented on qa/acceptance-tests/src/test/java/io/camunda/it/client/UsersByRoleIntegrationTest.java:
```diff
@@ -0,0 +1,430 @@
+    final var roleId = Strings.newRandomValidIdentityId();
+    createUser(username);
+    createRole(roleId, "roleName", "roleDesc");
+    camundaClient.newAssignUserToRoleCommand(roleId).username(username).send().join();
+
+    // when
+    camundaClient.newUnassignUserFromRoleCommand(roleId).username(username).send().join();
```
🔧 To ensure the test functions as it should be we need an await we assign the role to the user. Since secondary storage has a delay it could now happen that we assign the role to the user, don't actually unassign it and that the assertion still passes as secondary storage hasn't updated yet.

remcowesterhoud commented on qa/acceptance-tests/src/test/java/io/camunda/it/client/UsersByRoleIntegrationTest.java:
```diff
@@ -0,0 +1,430 @@
+  }
+
+  @Test
+  void shouldRejectUnassignIfRoleIsNull() {
+    assertThatThrownBy(
+            () ->
+                camundaClient
+                    .newUnassignUserFromRoleCommand(null)
+                    .username("username")
+                    .send()
+                    .join())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("roleId must not be null");
+  }
+
+  @Test
+  void shouldRejectUnassignIfRoleIsEmpty() {
+    assertThatThrownBy(
+            () ->
+                camundaClient.newUnassignUserFromRoleCommand("").username("username").send().join())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("roleId must not be empty");
+  }
+
+  @Test
+  void shouldRejectUnassignIfUsernameIsNull() {
+    assertThatThrownBy(
+            () ->
+                camundaClient.newUnassignUserFromRoleCommand("roleId").username(null).send().join())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("username must not be null");
+  }
+
+  @Test
+  void shouldRejectUnassignIfUsernameIsEmpty() {
+    assertThatThrownBy(
+            () -> camundaClient.newUnassignUserFromRoleCommand("roleId").username("").send().join())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("username must not be empty");
+  }
```
🔧 Let's remove this and make them part of the client unit tests instead.

remcowesterhoud commented on qa/acceptance-tests/src/test/java/io/camunda/it/client/UsersByRoleIntegrationTest.java:
```diff
@@ -0,0 +1,430 @@
+  }
+
+  @Test
+  void shouldRejectSearchIfRoleIdIsNull() {
+    assertThatThrownBy(() -> camundaClient.newUsersByRoleSearchRequest(null).send().join())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("roleId must not be null");
+  }
+
+  @Test
+  void shouldRejectSearchIfRoleIdIsEmpty() {
+    assertThatThrownBy(() -> camundaClient.newUsersByRoleSearchRequest("").send().join())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("roleId must not be empty");
+  }
```
🔧 Let's remove this and make them part of the client unit tests instead.

remcowesterhoud reviewed (changes requested):
Great work @Oleksiivanov. The main problem with the PR is that we need to rename some stuff, sorry about that 🙇 I didn't tag all places that need to change as it'd be way too much comments.

Furthermore some smaller comments regarding tests. Please take a look 👀

ana-vinogradova-camunda commented:
❓ should `RoleAuthorizationIT` be updated too?

remcowesterhoud commented:
> ❓ should `RoleAuthorizationIT` be updated too?

Would be nice, could be done in a different PR as well 🤷

Oleksiivanov commented:
@remcowesterhoud , could you please review this PR again?

I've made the following updates:
 - Rebased last main
 - Migrated UsersByRole to use RoleUser types 
 - Disabled UsersByRoleIntegrationTest for RDBMS
 - Added support for sorting by memberId in RoleFieldSortingTransformer
